components/script/script_module.rs

```diff
@@ -0,0 +1,541 @@
+    metadata: Option<Metadata>,
+    /// The initial URL requested.
+    url: ServoUrl,
+    /// Indicates whether the request failed, and why
+    status: Result<(), NetworkError>,
+    /// Timing object for this resource
+    resource_timing: ResourceFetchTiming,
+}
+
+impl FetchResponseListener for ModuleContext {
+    fn process_request_body(&mut self) {} // TODO(cybai): Perhaps add custom steps to perform fetch here?
+
+    fn process_request_eof(&mut self) {} // TODO(cybai): Perhaps add custom steps to perform fetch here?
+
+    fn process_response(&mut self, metadata: Result<FetchMetadata, NetworkError>) {
+        // https://html.spec.whatwg.org/multipage/#fetch-a-single-module-script
+        // Step 4.
+        let global = self.owner.global();
+        global.set_module_map(self.url.clone(), ModuleObject::Fetching);
+
+        self.metadata = metadata.ok().map(|meta| match meta {
+            FetchMetadata::Unfiltered(m) => m,
+            FetchMetadata::Filtered { unsafe_, .. } => unsafe_,
+        });
+
+        let status_code = self
+            .metadata
+            .as_ref()
+            .and_then(|m| match m.status {
+                Some((c, _)) => Some(c),
+                _ => None,
+            })
+            .unwrap_or(0);
+
+        self.status = match status_code {
+            0 => Err(NetworkError::Internal(
+                "No http status code received".to_owned(),
+            )),
+            200...299 => Ok(()), // HTTP ok status codes
+            _ => Err(NetworkError::Internal(format!(
+                "HTTP error code {}",
+                status_code
+            ))),
+        };
+    }
+
+    fn process_response_chunk(&mut self, mut chunk: Vec<u8>) {
+        if self.status.is_ok() {
+            self.data.append(&mut chunk);
+        }
+    }
+
+    /// <https://html.spec.whatwg.org/multipage/#fetch-a-single-module-script>
+    /// Step 8-14
+    fn process_response_eof(&mut self, response: Result<ResourceFetchTiming, NetworkError>) {
+        let global = self.owner.global();
+
+        let load = response.and(self.status.clone()).map(|_| {
+            let meta = self.metadata.take().unwrap();
+
+            let encoding = meta
+                .charset
+                .and_then(|encoding| Encoding::for_label(encoding.as_bytes()))
+                .unwrap_or(UTF_8);
+
+            // Step 12-1 & 13-1.
+            let (source_text, _, _) = encoding.decode(&self.data);
+            ScriptOrigin::external(
+                DOMString::from(source_text),
+                meta.final_url,
+                ScriptType::Module,
+            )
+        });
+
+        // TODO: HANDLE MIME TYPE CHECKING
+
+        match load {
+            Err(_) => {
+                // https://html.spec.whatwg.org/multipage/#fetch-a-single-module-script
+                // Step 9.
+                global.set_module_map(self.url.clone(), ModuleObject::Fetched(None));
+            },
+            Ok(resp_mod_script) => {
```

Comment:
Sorry for the messy commits again 😓

It belongs to the Step 11 & 12 [fetch-a-single-module-script](https://html.spec.whatwg.org/multipage/#fetch-a-single-module-script) (I wrote Step 14 because there was also JSON modules. But, supporting JSON modules has been removed from html spec. Will update the steps)

(I refactored them [here](https://github.com/servo/servo/pull/23545/files#diff-3fe97584f564214ec8e7ebbf91747e03R798-R843))